src/core/server/saved_objects/types.ts

```diff
@@ -487,3 +493,18 @@ export interface SavedObjectsTypeManagementDefinition<Attributes = any> {
 export type SavedObjectsExportablePredicate<Attributes = unknown> = (
   obj: SavedObject<Attributes>
 ) => boolean;
+
+/**
+ * If defined, allows a type to run a search query and return a query filter that may match any documents which may
+ * be excluded from the next migration upgrade process. Useful for cleaning up large numbers of old documents which
+ * are no longer needed and may slow the migration process.
+ *
+ * If this hook fails, the migration will proceed without these documents having been filtered out, so this
+ * should not be used as a guarantee that these documents have been deleted.
+ *
+ * @public
+ * @alpha Experimental and subject to change
+ */
+ export type SavedObjectTypeExcludeFromUpgradeFilterHook = (
+  esClient: Pick<ElasticsearchClient, 'search'>
```

Comment:
```suggestion
  { readonlyEsClient: Pick<ElasticsearchClient, 'search'> }
```
Hints the client is meant to read data.